lrc/Pos/Lrc/Fts.hs

```diff
@@ -1,52 +1,248 @@
 {-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TemplateHaskell     #-}
 
 -- | Base part of /follow-the-satoshi/ procedure.
 
 module Pos.Lrc.Fts
        ( followTheSatoshiM
        ) where
 
+import           Control.Lens       (makeLenses, makePrisms, uses)
 import           Data.List.NonEmpty (fromList)
 import           Universum
 
-import           Pos.Core.Coin      (coinToInteger, unsafeAddCoin)
+import           Pos.Core.Coin      (coinToInteger, unsafeGetCoin)
 import           Pos.Core.Constants (epochSlots)
-import           Pos.Core.Types     (Coin, SharedSeed (..), StakeholderId, mkCoin)
+import           Pos.Core.Types     (Coin, mkCoin, SharedSeed (..),
+                                     SlotLeaders, StakeholderId,
+                                     LocalSlotIndex(..))
+
 import           Pos.Crypto         (deterministic, randomNumber)
 import           Pos.Util.Iterator  (MonadIterator (..))
 
--- | A version of 'followTheSatoshi' that uses an iterator over 'TxOut's
--- instead of 'Utxo'.
+-- | Whereas 'Coin' stores an amount of coins, 'CoinIndex' is an identifier
+-- for a particular coin.
+--
+-- Here's an example where we assign indices to particular coins that belong
+-- to particular stakeholders:
+--
+-- StakeholderId | Coin (amount) | Coin indices (inclusive range)
+-- --------------+---------------+-----------------------------
+--           xxx | 200           |   [0..199]
+--           yyy | 50            | [200..249]
+--           zzz | 80            | [250..329]
+--
+-- In other words, 'Coin' is a cardinal, whereas 'CoinIndex' is an ordinal.
+newtype CoinIndex = CoinIndex Integer
+    deriving (Show, Eq, Ord, Num, Enum)
+
+makePrisms ''CoinIndex
+
+-- | Offset a coin index by a given amount of coins.
+coinIndexOffset :: Coin -> CoinIndex -> CoinIndex
+coinIndexOffset c = over _CoinIndex (+ fromIntegral (unsafeGetCoin c))
```

Comment:
@int-index are there any reasons not to use `unsafeAddCoin`? if there aren't, use it plz